jicofo/src/main/java/org/jitsi/jicofo/conference/JitsiMeetConferenceImpl.java

```diff
@@ -1137,78 +1137,49 @@ public void onTransportReject(@NotNull JingleSession jingleSession, @NotNull Jin
+     * participant is invalid).
      */
-    public StanzaError onAddSource(
-            @NotNull JingleSession jingleSession,
-            @NotNull List<? extends ContentPacketExtension> contents)
+    public void addSource(
+            @NotNull Participant participant,
+            @NotNull EndpointSourceSet sourcesAdvertised)
+    throws SenderCountExceededException, ValidationFailedException
     {
-        Jid address = jingleSession.getRemoteJid();
-        Participant participant = getParticipant(jingleSession);
-        if (participant == null)
-        {
-            String errorMsg = "no session for " + address;
-            logger.warn(errorMsg);
-            return StanzaError.from(StanzaError.Condition.item_not_found, errorMsg).build();
-        }
-
-        if (participant.getChatMember().getRole() == MemberRole.VISITOR)
-        {
-            return StanzaError.from(StanzaError.Condition.forbidden, "add-source not allowed for visitors").build();
-        }
-
-        String participantId = participant.getEndpointId();
-        EndpointSourceSet sourcesAdvertised = EndpointSourceSet.fromJingle(contents);
-        logger.debug(() -> "Received source-add from " + participantId + ": " + sourcesAdvertised);
-        if (sourcesAdvertised.isEmpty())
-        {
-            logger.warn("Received source-add with empty sources, ignoring");
-            return null;
-        }
-
         boolean rejectedAudioSource = sourcesAdvertised.getHasAudio() &&
                 chatRoom.getAudioSendersCount() >= ConferenceConfig.config.getMaxAudioSenders();
+        boolean rejectedVideoSource = sourcesAdvertised.getHasVideo() &&
+                chatRoom.getVideoSendersCount() >= ConferenceConfig.config.getMaxVideoSenders();
 
-        if (rejectedAudioSource ||
-                sourcesAdvertised.getHasVideo() &&
-                chatRoom.getVideoSendersCount() >= ConferenceConfig.config.getMaxVideoSenders())
-        {
-            String errorMsg = "Source add rejected. Maximum number of " +
-                    (rejectedAudioSource ? "audio" : "video") + " senders reached.";
-            logger.warn(() -> participantId + ": " + errorMsg);
-            return StanzaError.from(StanzaError.Condition.resource_constraint, errorMsg).build();
-        }
-
-        EndpointSourceSet sourcesAccepted;
-        try
+        if (rejectedAudioSource || rejectedVideoSource)
         {
-            sourcesAccepted = conferenceSources.tryToAdd(participant.getMucJid(), sourcesAdvertised);
-        }
-        catch (ValidationFailedException e)
-        {
-            logger.error("Error adding SSRCs from: " + address + ": " + e.getMessage());
-            return StanzaError.from(StanzaError.Condition.bad_request, e.getMessage()).build();
+            throw new SenderCountExceededException(
+                    "Sender count exceeded for: " + (rejectedAudioSource ? "audio " : "")
+                            + (rejectedVideoSource ? "video" : ""));
```

Comment:
If it's both, this'll say `exceeded for: audiovideo`.  Is that what you want?

Reply:
It should say `exceeded for: audio video`, because I added a space in the "audio" string. I thought that's a reasonable compromise with the unnecessarily verbose java code needed to make it nice (at least I couldn't think of a simple way to do it).